File: src/main/java/pcb/units/base/BaseUnit.java
package pcb.units.base;

import java.math.BigDecimal;
import java.util.Objects;
import java.util.function.Function;

public class BaseUnit<U extends Unit>
		implements Unit {

	// region private fields

	private final String symbol;
	private final String singularName;
	private final String pluralName;
	private final Function<BigDecimal, BigDecimal> translationToCanonical;
	private final Function<BigDecimal, BigDecimal> translationFromCanonical;

	// endregion

	// region constructors

	public BaseUnit(
			String symbol,
			String singularName,
			String pluralName,
			Function<BigDecimal, BigDecimal> translationToCanonical,
			Function<BigDecimal, BigDecimal> translationFromCanonical) {

		this.symbol = symbol;
		this.singularName = singularName;
		this.pluralName = pluralName;
		this.translationToCanonical = translationToCanonical;
		this.translationFromCanonical = translationFromCanonical;
	}

	// endregion

	// region implement Unit

	@Override
	public String getSymbol() {
		return symbol;
	}

	@Override
	public String getSingularName() {
		return singularName;
	}

	@Override
	public String getPluralName() {
		return pluralName;
	}

	@Override
	public Function<BigDecimal, BigDecimal> getTranslationToCanonical() {
		return translationToCanonical;
	}

	@Override
	public Function<BigDecimal, BigDecimal> getTranslationFromCanonical() {
		return translationFromCanonical;
	}

	// endregion

	// region overload Object

	@Override
	public boolean equals(Object obj) {
		if (obj instanceof Unit) {
			BaseUnit<?> other = (BaseUnit<?>) obj;

			return Objects.equals(this.symbol, other.symbol) &&
					Objects.equals(this.singularName, other.singularName) &&
					Objects.equals(this.pluralName, other.pluralName) &&
					Objects.equals(this.translationToCanonical, other.translationToCanonical) &&
					Objects.equals(this.translationFromCanonical, other.translationFromCanonical);
		}

		return false;
	}

	@Override
	public int hashCode() {
		return Objects.hash(symbol, singularName, pluralName, translationToCanonical, translationFromCanonical);
	}

	// endregion
}

File: dimension/src/main/java/pcb/uwu/core/BaseUnit.kt
package pcb.uwu.core

import pcb.uwu.core.Amount.Companion.ONE
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.unit.base.ScalarUnit
import pcb.uwu.util.hash

abstract class BaseUnit(private val precedence: Int,
                        override val symbol: String,
                        override val singularName: String,
                        override val pluralName: String,
                        override val translationToCanonical: (Amount) -> Amount,
                        override val translationFromCanonical: (Amount) -> Amount,
                        val magnitude: Magnitude = NATURAL)
    : Unit,
      Comparable<BaseUnit>
{
    /**
     * {@inheritDoc}
     */
    override val isScalar: Boolean
        get() = this is ScalarUnit

    // region Comparable

    override fun compareTo(other: BaseUnit): Int
    {
        return this.precedence - other.precedence
    }

    //endregion

    // region Object

    override fun equals(other: Any?): Boolean
    {
        if (this === other) return true
        if (other !is Unit) return false
        return symbol == other.symbol &&
                singularName == other.singularName &&
                pluralName == other.pluralName &&
                toCanonical(ONE) == other.toCanonical(ONE) &&
                fromCanonical(ONE) == other.fromCanonical(ONE)
    }

    override fun hashCode(): Int
    {
        return hash(symbol, singularName, pluralName, translationToCanonical, translationFromCanonical)
    }
}